protocol/contracts/nexus/Shards.sol

```diff
@@ -0,0 +1,232 @@
+    }
+
+    // called from Relic during Transmutations
+    function mintFromRelic(uint256 _shardId, uint256 _amount) external isRelic{
+        _mint(address(RELIC), _shardId, _amount,"");
+    }
+
+    // called from Relic during Transmutations
+    function burnFromRelic(uint256 _shardId, uint256 _amount) external isRelic{
+        _burn(address(RELIC), _shardId, _amount);
+    }
+
+     // use receipes to transform ingredients into a new item
+    function transmute(uint256 _recipeId)
+        external
+        nonReentrant
+    {
+        require(RELIC.balanceOf(msg.sender)>0, "You must have at least one Relic to transmute Shards");
+        
+        Recipe memory transmutation = recipes[_recipeId];
+        // Destroy
+        for (uint256 i = 0; i < transmutation.requiredIds.length; i++) {
+            require(
+                balanceOf(msg.sender, transmutation.requiredIds[i]) >=
+                    transmutation.requiredAmounts[i],
+                "Not enough ingredients"
+            );
+        }
+        _burnBatch(
+                msg.sender,
+                transmutation.requiredIds,
+                transmutation.requiredAmounts
+        );
+        // Create
+        
+        _mintBatch(
+                msg.sender,
+                transmutation.rewardIds,
+                transmutation.rewardAmounts,
+                ""
+            );      
+        emit Transmutation(msg.sender, _recipeId);
+    }
+
+     // For partners to mint their items, logic implemented on their side
+    function partnerMint(
+        address _to,
+        uint256 _id,
+        uint256 _amount,
+        bytes memory data
+    ) external {
+        require(partnerMinters[msg.sender], "You're not authorised to mint");
+        require(partnerAllowedIds[msg.sender][_id], "This isn't your reserved itemId");
+        _mint(_to, _id, _amount, data);
+        emit MintedShard(_to, _id, _amount);
+    }
+
+    function uri(uint256 _id) override public view returns(string memory){
+        return tokenURIs[_id];
+    }
+
+    //------- Owner -------//
+
+    // authorise a partner to mint 
+    function addPartner(address _partnerMinter, bool _flag) external onlyOwner {
+        partnerMinters[_partnerMinter]=_flag;
+    }
+    
+    // authorise specific shard ids to be minted by a partner
+    function whiteListItemsForPartner(address _partner, uint256[] memory _allowedIds, bool _allowed) external onlyOwner{
+        for(uint i = 0;i<_allowedIds.length;i++){
+            partnerAllowedIds[_partner][_allowedIds[i]]=_allowed;
+        }
+    }
+
+    function setRelic(address _relic) external onlyOwner {
+        RELIC = IRelic(_relic);
+    }
+
+    function createRecipe(
+        uint256 _recipeId,
+        uint256[] memory _requiredIds,
+        uint256[] memory _requiredAmounts,
+        uint256[] memory _rewardIds,
+        uint256[] memory _rewardAmounts
+    ) external onlyOwner {
+        recipes[_recipeId].id = uint16(_recipeId);
```

Comment:
Can also write this in one go with:
```
recipes[_recipeId] = Recipe({
  requiredIds: _requiredIds,
  requiredAmounts: _requiredAmounts,
  rewardIds: _rewardIds,
  rewardAmounts: _rewardAmounts
})
```

I would also recommend adding an event here.